### ISSUE TITLE
[feature] tf.contrib.image.invert

### ISSUE THREAD

**Issue by cancan101** [tensorflow/tensorflow#10737]
/CC @ringw 

Along the lines of the other utility functions in [`tf.contrib.image`](https://www.tensorflow.org/api_docs/python/tf/contrib/image), would be great to have a `tf.contrib.image.invert` that takes a transform and returns the inverse transform.

**Comment by aselle** [comment]
Could you give an example of how this would be used? Also, how would you imagine implementing this. It seems to be it would be difficult, because you'd have to introspect a general tensor to make it work like you suggest.

**Comment by cancan101** [comment]
Not sure what you mean about introspecting tensor. I would think this would just convert the transform, represented as a tensor of shape (num_images, 8), to the standard matrix representation of an affine transform: `num_images x 3 x 3`, and then doing a standard matrix inverse.

For some more info:
* http://scikit-image.org/docs/dev/api/skimage.transform.html#affinetransform

This would be used in the case where you want to take an input image, transform it, run that transformed image through a network, get a result (may be coordinates or an output image) and then transform the results back to the original space.

**Comment by martinwicke** [comment]
I don't think this is feasible in general. I'll close this issue. If someone can come up with a workable design, please file a new issue.

**Comment by cancan101** [comment]
Maybe I confused matters. This should just be a matter of reshaping the 8 values to 9 and then calling [`tf.matrix_inverse`](https://www.tensorflow.org/api_docs/python/tf/matrix_inverse).

**Comment by cancan101** [comment]
In fact that reshaping method exists: [`_flat_transforms_to_matrices`](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/contrib/image/python/ops/image_ops.py#L201-L209) it just isn't exported.

**Comment by ringw** [comment]
This also seems feasible to me.

**Comment by cancan101** [comment]
@ringw what do you think about exposing `_flat_transforms_to_matrices` ?

**Comment by ringw** [comment]
That's fine with me, if we also expose `_transform_matrices_to_flat` for consistency.

**Comment by aselle** [comment]
This sounds fine. @cancan101, I was confused that you meant "Transforms" in the general sense ... where contrast and brightness are a type of "transform" and you wanted to be able to create an op that undoes things. Would you be willing to submit a pull request @cancan101.

**Comment by 3rd3** [comment]
I've run into numerical instabilities inverting transformation matrices before, though for data augmentation it should suffice. Usually one can construct the same matrices again using their inverse primitive affine transformations (shear, rotation etc.). Perhaps a function would be helpful that either constructs the mapping or the inverse like this `tf.image.affine_transform(shear_x, shear_y, rotation, scale, inverse=False)`. Not entirely sure how useful that is. Just an idea.

**Comment by ringw** [comment]
I'd argue against adding an `inverse` argument for rotation and translation
where the arguments can just be negated, but if it's useful to have a
`shear` function, we might want an inverse argument there. (Any function
like `rotate` should have a corresponding function that creates the
transform, and then the user can compose the transforms in whatever order
they want.)

On Jun 18, 2017 12:01 AM, "3rd3" <notifications@github.com> wrote:

I've run into numerical instabilities inverting transformation matrices
before, though for data augmentation it should suffice. Usually one can
construct the same matrices again using their inverse primitive affine
transformations (shear, rotation etc.). Perhaps a function would be helpful
that either constructs the mapping or the inverse like this
tf.image.affine_transform(shear_x, shear_y, rotation, scale, inverse=False).
Not entirely sure how useful that is. Just an idea.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/10737#issuecomment-309242360>,
or mute the thread

**Comment by dynamicwebpaige** [comment]
Though supporting `tf.image` inversions is outside of the scope of TensorFlow Core, it might be a welcome contribution to image processing in [TensorFlow Addons](https://github.com/tensorflow/addons/tree/master/tensorflow_addons/image).

Transferring this issue to the `tensorflow/addons` repo; @seanpmorgan for visibility.
